### ISSUE TITLE
Incorrect GB speed/sound pitch with audio distortion in Super GameBoy 2 emulation [bsnes 112.x]

### ISSUE THREAD

**Issue by richterdeveloper** [byuu/bsnes#209]
In SGB 2 the speed/audio pitch of the GB game is the same as the SGB 1 (in real hardware, the SGB 2 fixed that issue present in the original SGB). In addition, the sound of the GB is distorted from time to time . The SGB 1 emulation is working properly.

Tested using the No-Intro SGB roms:

Super Game Boy 2 (Japan).sfc  
- sha1: E5B2922CA137051059E4269B236D07A22C07BC84

See the video at 0:57 (using the SGB 2 rom)
https://streamable.com/czds3

**Comment by byuu** [comment]
@LIJI32 any chance you can take a look at this one? I've done all I could reasonably do with getting SGB audio to sound good. SGB2 sounded good to me, but I'm clearly not an audiophile when it comes to this sort of thing...

**Comment by LIJI32** [comment]
Newer (master) versions of SameBoy added the `GB_set_sample_rate_by_clocks` API which should fix all these weird bsnes-SameBoy audio integration issues. Assuming 48KHz output, it should be called with:

* `0x400000 / 48000 = 87.38133333...` for SGB2
* `21477272 / 5 / 48000 = 89.4886333...` for SGB on an NTSC SNES
* `21281370 / 5 / 48000 = 88.672375` for SGB on a PAL SNES

And while not THAT important, it's generally a good idea to use `GB_MODEL_SGB_NTSC_NO_SFC`, `GB_MODEL_SGB_PAL_NO_SFC` and `GB_MODEL_SGB2_NO_SFC` when initializing SameBoy so it consistently emulates analog artifacts not affected by the clock rate (Such as DAC decays).

Let me know if it helped.

**Comment by byuu** [comment]
@LIJI32 I merged the latest SameBoy master (2019-11-05)
I re-added GB_set_read_memory_callback, but made it use the new gb->read_memory_callback. I did not test cheat codes with the nwe core yet.
Unfortunately, inputs are busted in SGB mode now. I can't make Pokemon Yellow do anything. Because of that, I'm keeping the old core around, meaning I can't test the new GB_set_sample_rate_by_clocks API until that's resolved.
As always, thanks for your help with this.
